src/odemis/gui/comp/text.py

```diff
@@ -618,6 +618,78 @@ def _step_from_range(min_val, max_val):
         logging.exception(msg)
 
 
+class PatternValidator(ValidatorClass):
+
+    def __init__(self, pattern):
+        """ pattern (str): regex pattern of allowed entries """
+        super().__init__()
+        self.Bind(wx.EVT_CHAR, self.on_char)
+        self.entry_pattern = pattern
+
+    def _is_valid_value(self, val):
+        """ Validate the given value
+
+        Args:
+            val (str):
+
+        Returns:
+            (boolean): True if the given string is valid
+
+        """
+
+        entry_pattern = re.compile(self.entry_pattern)
+        return bool(re.match(entry_pattern, val)) and entry_pattern.match(val).end() == len(val)
+
+    def _get_str_value(self):
+        """ Return the string value of the wx.Window to which this validator belongs """
+
+        # Special trick in case we are validating a NumberTextCtrl, which has it's
+        # default 'GetValue' method replaced with one that returns number instances
+
+        fld = self.GetWindow()
+
+        if hasattr(fld, "get_value_str"):
+            val = fld.get_value_str()
+        else:
+            val = fld.GetValue()
+        return val
+
+    def on_char(self, event):
+        """ This method prevents the entry of illegal characters """
+        ukey = event.GetUnicodeKey()
+        # Allow control keys to propagate (most of them are WXK_NONE in unicode)
+        if ukey < wx.WXK_SPACE:
+            event.Skip()
+            return
+
+        field_val = str(self._get_str_value())
+        start, end = self.GetWindow().GetSelection()
+        field_val = field_val[:start] + chr(ukey) + field_val[end:]
+
+        entry_pattern = re.compile(self.entry_pattern)
```

Comment:
You shouldn't compile the regex everytime you use it. Compile it at init, then use the pattern.